File: apps/client/src/lib/github-oauth.ts
import { invoke } from "@tauri-apps/api/core";

export interface GitHubOAuthConfig {
  server_url: string;
}

export interface AuthUrlRequest {
  public_key: any;
  username: string;
}

export interface AuthUrlResponse {
  auth_url: string;
  state: string;
}

export interface GitHubIdentityRequest {
  code: string;
  state: string;
  username: string;
  challenge_signature: string;
}

export interface GitHubIdentityResponse {
  identity_pod: any;
}

export interface GitHubServerInfo {
  server_id: string;
  public_key: any;
}

/**
 * GitHub OAuth API client for POD2 identity server integration
 */
export class GitHubOAuthClient {
  private serverUrl: string;

  constructor(config: GitHubOAuthConfig) {
    this.serverUrl = config.server_url;
  }

  /**
   * Get server information from GitHub OAuth identity server
   */
  async getServerInfo(): Promise<GitHubServerInfo> {
    const response = await fetch(this.serverUrl);
    if (!response.ok) {
      throw new Error(`Failed to get server info: ${response.status}`);
    }
    return response.json();
  }

  /**
   * Get GitHub OAuth authorization URL
   */
  async getAuthUrl(publicKey: any, username: string): Promise<AuthUrlResponse> {
    const response = await fetch(`${this.serverUrl}/auth/github`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        public_key: publicKey,
        username: username,
      }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Failed to get auth URL: ${response.status} - ${errorText}`);
    }

    return response.json();
  }

  /**
   * Complete GitHub OAuth identity verification
   */
  async completeIdentityVerification(request: GitHubIdentityRequest): Promise<GitHubIdentityResponse> {
    const response = await fetch(`${this.serverUrl}/identity`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Failed to complete identity verification: ${response.status} - ${errorText}`);
    }

    return response.json();
  }

  /**
   * Open GitHub OAuth URL in external browser
   */
  async openAuthUrl(authUrl: string): Promise<void> {
    const { openUrl } = await import('@tauri-apps/plugin-opener');
    await openUrl(authUrl);
  }
}

/**
 * Enhanced identity server setup that detects GitHub OAuth servers
 */
export async function setupGitHubIdentityServer(serverUrl: string): Promise<{
  server_info: GitHubServerInfo;
  is_github_server: boolean;
}> {
  // Use Tauri backend to detect GitHub OAuth server
  const is_github_server = await invoke('detect_github_oauth_server', { serverUrl }) as boolean;
  
  // Get server info
  const client = new GitHubOAuthClient({ server_url: serverUrl });
  const server_info = await client.getServerInfo();
  
  return { server_info, is_github_server };
}

/**
 * Register username with GitHub OAuth flow
 */
export async function registerWithGitHubOAuth(
  serverUrl: string,
  username: string
): Promise<{
  identity_pod: any;
  username: string;
  server_id: string;
}> {
  // Get user's public key from Tauri backend
  const privateKeyInfo = await invoke('get_private_key_info') as { public_key: any };
  const publicKey = privateKeyInfo.public_key;

  const client = new GitHubOAuthClient({ server_url: serverUrl });

  // Step 1: Get GitHub OAuth authorization URL
  const authResponse = await client.getAuthUrl(publicKey, username);

  // Step 2: Open GitHub OAuth URL in browser
  await client.openAuthUrl(authResponse.auth_url);

  // Step 3: Wait for user to complete OAuth and return authorization code
  // This would typically be handled by a callback or user input
  // For now, we'll return a promise that resolves when the user provides the code
  return new Promise((_resolve, reject) => {
    // This would be implemented with a proper OAuth callback handler
    // For now, we'll just throw an error indicating this needs to be implemented
    reject(new Error('OAuth callback handling not yet implemented. User needs to provide authorization code manually.'));
  });
}

/**
 * Extract GitHub-specific information from identity POD
 */
export function extractGitHubInfoFromIdentityPod(identityPod: any): {
  username: string;
  github_username?: string;
  github_user_id?: number;
  github_public_keys?: string[];
  github_email?: string;
  oauth_verified_at?: string;
} {
  const username = identityPod.username || 'Unknown';
  
  // Parse GitHub data dictionary
  let githubData: any = {};
  if (identityPod.github_data) {
    try {
      githubData = typeof identityPod.github_data === 'string' 
        ? JSON.parse(identityPod.github_data) 
        : identityPod.github_data;
    } catch (error) {
      console.error('Failed to parse GitHub data from identity POD:', error);
    }
  }

  return {
    username: username,
    github_username: githubData.github_username,
    github_user_id: githubData.github_user_id,
    github_public_keys: githubData.github_public_keys,
    github_email: githubData.github_email,
    oauth_verified_at: githubData.oauth_verified_at,
  };
}

File: apps/client/src-tauri/src/features/identity_setup/github_oauth.rs
use anyhow::Result;
use pod2::middleware::TypedValue;
use serde::{Deserialize, Serialize};
use tauri::State;
use tokio::sync::Mutex;

use crate::AppState;

#[derive(Debug, Serialize, Deserialize)]
pub struct GitHubAuthUrlRequest {
    pub public_key: serde_json::Value,
    pub username: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GitHubAuthUrlResponse {
    pub auth_url: String,
    pub state: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GitHubIdentityRequest {
    pub code: String,
    pub state: String,
    pub username: String,
    pub challenge_signature: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GitHubIdentityResponse {
    pub identity_pod: serde_json::Value,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GitHubIdentityPodResult {
    pub identity_pod: serde_json::Value,
    pub username: String,
    pub github_username: Option<String>,
    pub server_id: String,
}

/// Get GitHub OAuth authorization URL
#[tauri::command]
pub async fn get_github_auth_url(
    server_url: String,
    username: String,
    state: State<'_, Mutex<AppState>>,
) -> Result<GitHubAuthUrlResponse, String> {
    log::info!("Getting GitHub OAuth authorization URL for user: {username}");

    // Get or create the user's private key during setup (same as regular identity setup)
    let app_state = state.lock().await;
    let private_key = match pod2_db::store::get_default_private_key_raw(&app_state.db).await {
        Ok(key) => {
            log::info!("Using existing default private key for GitHub OAuth");
            key
        }
        Err(_) => {
            log::info!("Creating new default private key for GitHub OAuth");
            pod2_db::store::create_default_private_key(&app_state.db)
                .await
                .map_err(|e| format!("Failed to create private key: {e}"))?
        }
    };

    let public_key = private_key.public_key();
    drop(app_state); // Release the lock before making HTTP requests

    let client = reqwest::Client::new();
    let request = GitHubAuthUrlRequest {
        public_key: serde_json::to_value(public_key)
            .map_err(|e| format!("Failed to serialize public key: {e}"))?,
        username: username.clone(),
    };

    let response = client
        .post(format!("{server_url}/auth/github"))
        .json(&request)
        .send()
        .await
        .map_err(|e| format!("Failed to request GitHub auth URL: {e}"))?;

    if !response.status().is_success() {
        let status = response.status();
        let error_text = response.text().await.unwrap_or_default();
        return Err(format!(
            "GitHub auth URL request failed: {status} - {error_text}"
        ));
    }

    let auth_response: GitHubAuthUrlResponse = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse GitHub auth URL response: {e}"))?;

    log::info!("Successfully obtained GitHub auth URL for user: {username}");
    Ok(auth_response)
}

/// Complete GitHub OAuth identity verification
#[tauri::command]
pub async fn complete_github_identity_verification(
    server_url: String,
    code: String,
    state: String,
    username: String,
    app_state: State<'_, Mutex<AppState>>,
) -> Result<GitHubIdentityPodResult, String> {
    log::info!("Completing GitHub OAuth identity verification for user: {username}");

    // Get or create the user's private key during setup
    let mut state_lock = app_state.lock().await;
    let _private_key = match pod2_db::store::get_default_private_key_raw(&state_lock.db).await {
        Ok(key) => {
            log::info!("Using existing default private key");
            key
        }
        Err(_) => {
            log::info!("Creating new default private key");
            pod2_db::store::create_default_private_key(&state_lock.db)
                .await
                .map_err(|e| format!("Failed to create private key: {e}"))?
        }
    };

    // For now, we'll use a placeholder challenge signature
    // In a full implementation, this would involve proper challenge signing
    let challenge_signature = "placeholder_signature".to_string();

    let client = reqwest::Client::new();
    let identity_request = GitHubIdentityRequest {
        code,
        state,
        username: username.clone(),
        challenge_signature,
    };

    let response = client
        .post(format!("{server_url}/identity"))
        .json(&identity_request)
        .send()
        .await
        .map_err(|e| format!("Failed to complete GitHub identity verification: {e}"))?;

    if !response.status().is_success() {
        let status = response.status();
        let error_text = response.text().await.unwrap_or_default();
        return Err(format!(
            "GitHub identity verification failed: {status} - {error_text}"
        ));
    }

    let identity_response: GitHubIdentityResponse = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse GitHub identity response: {e}"))?;

    // Convert the identity POD from JSON to the actual SignedPod type
    let identity_pod: pod2::frontend::SignedPod =
        serde_json::from_value(identity_response.identity_pod.clone())
            .map_err(|e| format!("Failed to deserialize identity POD: {e}"))?;

    // Extract GitHub username from the identity POD's github_data dictionary
    let github_username = identity_pod
        .get("github_data")
        .and_then(|v| match v.typed() {
            TypedValue::String(s) => {
                // Parse the GitHub data JSON
                if let Ok(github_data) = serde_json::from_str::<serde_json::Value>(s.as_str()) {
                    github_data
                        .get("github_username")
                        .and_then(|u| u.as_str())
                        .map(|u| u.to_string())
                } else {
                    None
                }
            }
            _ => None,
        });

    // Store identity POD in database as mandatory
    let pod_data = pod2_db::store::PodData::Signed(Box::new(identity_pod.clone().into()));
    let identity_pod_id = pod_data.id();

    // Ensure "identity" folder exists
    const IDENTITY_FOLDER: &str = "identity";
    if !pod2_db::store::space_exists(&state_lock.db, IDENTITY_FOLDER)
        .await
        .unwrap_or(false)
    {
        pod2_db::store::create_space(&state_lock.db, IDENTITY_FOLDER)
            .await
            .map_err(|e| format!("Failed to create identity folder: {e}"))?;
        log::info!("✓ Created identity folder");
    }

    pod2_db::store::store_identity_pod(
        &state_lock.db,
        &pod_data,
        IDENTITY_FOLDER,
        Some("GitHub Identity POD"),
    )
    .await
    .map_err(|e| format!("Failed to store identity POD: {e}"))?;

    // Update setup state with username and identity POD ID
    pod2_db::store::update_identity_info(&state_lock.db, &username, &identity_pod_id)
        .await
        .map_err(|e| format!("Failed to update identity info: {e}"))?;

    // Trigger state sync to refresh UI with new identity POD
    state_lock
        .trigger_state_sync()
        .await
        .map_err(|e| format!("Failed to trigger state sync: {e}"))?;

    log::info!("Successfully completed GitHub OAuth identity verification for user: {username}");

    Ok(GitHubIdentityPodResult {
        identity_pod: identity_response.identity_pod,
        username,
        github_username,
        server_id: "github-identity-server".to_string(),
    })
}

/// Detect if a server is a GitHub OAuth identity server
#[tauri::command]
pub async fn detect_github_oauth_server(
    server_url: String,
    _state: State<'_, Mutex<AppState>>,
) -> Result<bool, String> {
    log::info!("Detecting if server is GitHub OAuth server: {server_url}");

    let client = reqwest::Client::new();
    let response = client
        .get(&server_url)
        .send()
        .await
        .map_err(|e| format!("Failed to connect to server: {e}"))?;

    if !response.status().is_success() {
        return Ok(false);
    }

    let server_info: serde_json::Value = response
        .json()
        .await
        .map_err(|_| "Failed to parse server response".to_string())?;

    // Check if server_id indicates GitHub OAuth server
    let is_github_server = server_info
        .get("server_id")
        .and_then(|v| v.as_str())
        .map(|id| id == "github-identity-server")
        .unwrap_or(false);

    log::info!("Server GitHub OAuth detection result: {is_github_server}");
    Ok(is_github_server)
}
